### PR TITLE
feat(ref-imp): #766 - Implemented hashing of public key as reveal value

### PR DESCRIPTION
**Pull request by thehenrytsai**

### CODE REVIEW

File: tests/fixtures/create/create.json
Comment (troyronda) [inline tests/fixtures/create/create.json:3]: Was there a corresponding change to the spec that has resulted in the fixtures changing?

Comment (csuwildcat) [reply tests/fixtures/create/create.json:3]: This is part of SIP1, the thing where we need to remove the revealed public key JWK value from the Anchor File, which requires hashing it twice, do you don't have to put the actual data there.

**Review by OR13 — DISMISSED** [review]
Changes to protocol markdown that is noted to be deprecated mixed with breaking change to implementation and test vectors.

"Double Hash" not described in spec.

Why do we need a "double hash"? This behavior should be documented. Before being merged in a way that breaks interoperability.

**Comment by csuwildcat** [comment]
@OR13 kind of a tough spot here, because this is the first thing that needs to be done to address SIP 1, and the rest of SIP 1 is about a month out. So we can either reflect this one piece in the spec in some intermediary state, or update the spec to include all of SIP 1 now. I would hate to dump a bunch of time into updating this if it's an intermediary version of the spec that should not really exist.

**Comment by csuwildcat** [comment]
A better way forward would be to have the latest spec simply state that it is not yet fully reflective of the updates in master of the reference. I can get started on all the SIP 1 changes to the spec, but I don't think we should blog all check-ins until all of SIP 1's facets are completed in both.

**Comment by thehenrytsai** [comment]
> Why do we need a "double hash"? This behavior should be documented. Before being merged in a way that breaks interoperability.

I will let @csuwildcat comment on documentation. But essentially, this is the first stage of implementing SIP1. This change front loads the changes needed by the client side to perform double hash scheme to get the commit value as specified by SIP1, without needing to wait until server-side SIP1 is implemented.

File: tests/fixtures/create/create.json
Comment (thehenrytsai) [reply tests/fixtures/create/create.json:3]: Adding to @csuwildcat, this does impact the fixture/vector because with this PR commit value needs to be double hash of public key as opposed to single hash.

**Comment by troyronda** [comment]
@csuwildcat @thehenrytsai It would be helpful to have some context on why SIP1 changes to double hash. Apologies if this was already explained elsewhere.

**Comment by OR13** [comment]
@csuwildcat @thehenrytsai there is nothing wrong with using a branch called SIP-1 and merging it into master when its been fully documented....

Pushing incremental breaking changes to master, especially ones that impact test vectors used for interoperability is a really bad signal to the market about the stability of sidetree.

I suggest you use a branch other than master, fully document the changes that will be merged into the reference implementation via the 'spec' and then open a PR which merges the breaking changes along with the necessary documentation changes into master.

You can publish pre-release packages from branches other than master.

master should be stable and usable, and not contain partial implementations of new features... thats what feature branches are for :)

**Comment by csuwildcat** [comment]
Another option is to create a branch of this state tagged with the current spec number and let master be master.

**Comment by csuwildcat** [comment]
@troyronda we talked about the reason for double hashing the revealed jwk extensively on last week's Sidetree call, but I will take a few minutes on today's call to do so again.

**Comment by OR13** [comment]
> Another option is to create a branch of this state tagged with the current spec number and let master be master.

https://thenewstack.io/dont-mess-with-the-master-working-with-branches-in-git-and-github/#:~:text=The%20master%20branch%20is%20meant,to%20work%20from%20the%20master.

File: lib/core/versions/latest/Multihash.ts
Comment (isaacJChen) [inline lib/core/versions/latest/Multihash.ts:175]: return Buffer.compare(actualMultihashBuffer, expectedMultihashBuffer) == 0

File: lib/core/versions/latest/Multihash.ts
Comment (isaacJChen) [inline lib/core/versions/latest/Multihash.ts:197]: Same simplification as above.

**Comment by thehenrytsai** [comment]
Hey @OR13,

In the WG meeting, the WG have:

1. Approved SIP1.
1. Approved this PR to be merged in to allow downstream clients/services to immediately start generating SIP1 compatible DIDs going forward without waiting on rest of the SIP1 specification/implementation.
1. Agreed that SIP1 spec must be immediately be worked on, finalized in the next week. A separate spec PR will be sent out.
1. Agreed that remaining/actual SIP1 implementation will follow after the SIP1 spec is checked in.

Thanks,
Henry = ]

**Comment by OR13** [comment]
Master should be marked as unstable for this repo, I remain opposed to incrementally merging partial implementation of features into master, especially if they are wildly out of sync with the spec... it makes everyones jobs harder.... 

However, as I am no long a chair / editor of the wg I cannot block any of this, I am just giving you my opinion, as a developer who has to maintain interop with this repo.

The CODEOWNERS for this repo, if there where any, could merge over my objection... I suggest someone set that up... here is an example from the W3C CCG https://github.com/w3c-ccg/universal-wallet-interop-spec/blob/master/CODEOWNERS

**Review by isaacJChen — APPROVED** [review]

**Comment by thehenrytsai** [comment]
@OR13, filed #844 to add CODEOWNERS file in a separate PR. Agree with the point about not merging in broken code/features into `master` in principle, and I believe most of us are following that principle. In this particular case, since SIP1 is officially approved, the WG agreed that it is beneficial for implementer to be aware and start generating Sidetree v1.0 compatible DIDs ASAP, so we agree to check in this PR is on its own. Your point on committed spec lagging behind implementation is taken, as such a PR for SIP1 will be sent out in a week's time or so, SIP1 implementation PR will follow.

The `master` is considered unstable by default currently, in fact Tobias updated the pipeline such that an unstable release is carved out every time a merge is made into master (unstable version can be seen on repo landing page). If you have specific steps to make this more clear, please let me know, I am certainly not as well-versed as you when it comes to github repo/branch management, so could use your help there! Not to say this paradigm won't change as we stabilize and approach spec v1.0, that is certainly the goal.

With above said, I will merge this PR in.

**Pull request merged**
